front: drop trainIdsToFetch state

emersion opened the pull request:
We don't actually need a state here: all setTrainIdsToFetch() calls were overwriting any previously stored value.

The state makes the whole thing harder to reason about, because a state update will (asynchronously) invoke any useEffect() that depends on that state, so it's harder to figure out what exact side effects will be caused by setting the state. A simple callback is easier to follow.

No behavior change intended.

anisometropie commented on front/src/applications/operationalStudies/hooks/useLazyLoadTrains.ts:
```diff
@@ -66,51 +58,44 @@ const useLazyLoadTrains = ({
-
-    if (infraId && trainIdsToFetch && trainIdsToFetch.length > 0) {
-      getTrainScheduleSummaries(infraId, trainIdsToFetch);
+  const loadTrainIds = async (trainIds: number[]) => {
+    setAllTrainsLoaded(false);
+
```
if we have no scenario we don’t get an infraId from `useScenarioData`. 
in that case we need not do anything and just return early.

```suggestion
  const loadTrainIds = async (trainIds: number[]) => {
    if (!infraId) {
      return;
    }

emersion replied:
Is this something which can really happen? Can we have a list of trains to load from a timetable without an infra?

If it can't happen, I'd rather fail hard if we hit this case, so that we can debug the root cause instead of silently doing nothing.

anisometropie replied:
alright, I think I’m okay to consider we should have a studyId defined at this point

anisometropie commented on front/src/applications/operationalStudies/hooks/useLazyLoadTrains.ts:
```diff
@@ -66,51 +58,44 @@ const useLazyLoadTrains = ({
+
+      const rawSummaries = await postV2TrainScheduleSimulationSummary({
+        body: {
+          infra_id: infraId!,
```
```suggestion
          infra_id: infraId,
```

clarani reviewed (commented):
When I try to add trains in a timetable which already have some, the new trains are not refetched :/

emersion committed:
front: drop trainIdsToFetch state

We don't actually need a state here: all setTrainIdsToFetch() calls
were overwriting any previously stored value.

The state makes the whole thing harder to reason about, because a
state update will (asynchronously) invoke any useEffect() that
depends on that state, so it's harder to figure out what exact side
effects will be caused by setting the state. A simple callback is
easier to follow.

No behavior change intended.

emersion committed:
front: simplify setProjectedTrainsById() call

No need for an intermediate variable here.

emersion committed:
front: simplify reset in useLazyProjectTrains()

emersion commented:
There is a race between the summaries endpoint and the train schedule endpoint. If `trainScheduleSummaries` is populated with the new data before `trainSchedules` is updated, `formatTrainScheduleSummaries()` will not include the new train.

emersion committed:
front: pass train schedules as argument to loadTrainIds()

There is a race between the /train_schedule endpoint (used to
retrieve all train schedules) and the /train_schedule/simulation_summary
endpoint (used to retrieve a batch of simulation summaries).
Depending on the order in which responses are received, the train
might be dropped. If /train_schedule/simulation_summary comes first,
formatTrainScheduleSummaries() will not have the train schedule
to work with, thus will ignore the train.

Fix this by passing a list of train schedules to loadTrainIds().
The function can retrieve simulation summaries and directly has
up-to-date train schedules to work with.

Rename the function to loadTrains() since it no longer takes IDs.

emersion commented:
I've fixed the race. It effectively reverts 6d020d86930318a09b9ca7b0d947ce1678055e65.

emersion committed:
front: move trainIdsToProject state into useLazyProjectTrains()

The list of trains that need to be projected is internal state
that doesn't need to exposed outside of useLazyProjectTrains().
Outsiders only need to request to project more trains. Expose a
function to do so instead.

emersion commented:
Couldn't resist the urge to add in one more refactor commit.

anisometropie reviewed (commented):
unfortunately it breaks the fix I did. The assert fails again in `isScheduledPointsNotHonored`, when you modify a train by adding a path step

![Screenshot_20240910_104846](https://github.com/user-attachments/assets/e881f52b-9bc9-4cfa-9aab-3ac8126b03a5)

emersion commented:
Superseded by https://github.com/OpenRailAssociation/osrd/pull/11390
